feat(ipns): records with V2-only signatures

hacdias opened the pull request:
See https://github.com/ipfs/boxo/pull/339.

- Note that this also removes the REFRAME routing support. It has been deprecated and the deprecated package has conflicts with the changes we made to `boxo/ipns`. It wouldn't make sense to update it since it's deprecated (https://github.com/ipfs/kubo/issues/9823#issuecomment-1572198887).

hacdias commented:
See https://github.com/ipfs/boxo/pull/339#issuecomment-1586927027

hacdias commented on core/commands/name/publish.go:
```diff
@@ -128,23 +128,17 @@ Alternatively, publish an <ipfs-path> using a valid PeerID (as listed by
 		return cmds.EmitOnce(res, &IpnsEntry{
-			Name:  keyEnc.FormatID(pid),
-			Value: out.Value().String(),
+			Name:  strings.TrimPrefix(name.String(), "/ipns/"),
```
I did not want to keep this. But I'm keeping for two reasons:
1. `ipns name publish` command keeps outputting the same expected output with just the Peer ID (note that it is now normalized).
2. interop with js-ipfs pubsub ipns tests fail if we change this because it uses the old RPC API

We can fix 1 and 2 though if that's desirable.

lidel replied:
Hm..  am kinda confused why we ended up with prefix as part of Name string. 
Is it because it was mentioned in https://specs.ipfs.tech/ipns/ipns-record/#ipns-name? 
Lets remove it – see https://github.com/ipfs/boxo/pull/339#discussion_r1229847123

lidel commented on core/commands/name/publish.go:
```diff
@@ -28,6 +28,7 @@ const (
 	ttlOptionName          = "ttl"
 	keyOptionName          = "key"
 	quieterOptionName      = "quieter"
+	compatibleWithV1       = "compatible-v1"
```
```suggestion
	v1compatOptionName       = "v1compat"
```

lidel commented on core/commands/name/publish.go:
```diff
@@ -83,19 +84,17 @@ Alternatively, publish an <ipfs-path> using a valid PeerID (as listed by
 		cmds.StringOption(ttlOptionName, "Time duration this record should be cached for. Uses the same syntax as the lifetime option. (caution: experimental)"),
 		cmds.StringOption(keyOptionName, "k", "Name of the key to be used or a valid PeerID, as listed by 'ipfs key list -l'.").WithDefault("self"),
 		cmds.BoolOption(quieterOptionName, "Q", "Write only final hash."),
+		cmds.BoolOption(compatibleWithV1, "Create a V1-compatible IPNS Record.").WithDefault(true),
```
```suggestion
		cmds.BoolOption(v1compatOptionName, "Produce a backward-compatible IPNS Record by including fields for both V1 and V2 signatures.").WithDefault(true),
```

lidel reviewed (changes requested):
Thanks! Lgtm, but needs tests, probably in `test/sharness/t0100-name.sh` (fine to add or migrate them to `test/cli/harness` if you dont want to write bash)

- [ ] `ipfs name publish`  + `resolve`  + `inspect` with and without `v1compat`

lidel commented on core/commands/name/publish.go:
```diff
@@ -83,19 +83,17 @@ Alternatively, publish an <ipfs-path> using a valid PeerID (as listed by
 		cmds.StringOption(ttlOptionName, "Time duration this record should be cached for. Uses the same syntax as the lifetime option. (caution: experimental)"),
 		cmds.StringOption(keyOptionName, "k", "Name of the key to be used or a valid PeerID, as listed by 'ipfs key list -l'.").WithDefault("self"),
 		cmds.BoolOption(quieterOptionName, "Q", "Write only final hash."),
+		cmds.BoolOption(v1compatOptionName, "Produce a backward-compatible IPNS Record by including fields for both V1 and V2 signatures.").WithDefault(true),
```
:thought_balloon:  This PR should have  information about this + related IPIP (when we create it  in `/docs/changelogs/v0.22.md` (ok to create a placeholder TODO in markdown for now, we can fill it up before 0.22  release)

lidel commented on docs/changelogs/v0.22.md:
```diff
@@ -6,13 +6,23 @@
 
 ### 🔦 Highlights
 
+#### `ipfs name publish` now supports V2 only IPNS records
+
+When publishing an IPNS record, you are now able to create v2 only records
+by passing `--v1compat=false`. By default, we still create V1+V2 records, such
+that there is the highest chance of backwards compatibility. The goal is to move
+to V2 only in the future.
+
+**TODO**: add links to IPIP.
```
```suggestion
#### `ipfs name publish` now supports V2 only IPNS records

When publishing an IPNS record, you are now able to create v2 only records
by passing `--v1compat=false`. By default, we still create V1+V2 records, such
that there is the highest chance of backwards compatibility. The goal is to move
to V2 only in the future.

**TODO**: add links to IPIP https://github.com/ipfs/specs/issues/376
```

lidel commented on core/commands/name/name.go:
```diff
@@ -151,78 +145,76 @@ Passing --verify will verify signature against provided public key.
+			return err
+		}
+		if pbRecord.SignatureV1 != nil || pbRecord.Value != nil {
+			if pbRecord.Data != nil {
+				result.Version = "V1+V2"
+			} else {
+				result.Version = "V2"
+			}
```
:thought_balloon: Is this a bug? If the Data is missing, it can't be V2. 
Seems we miss test for this.

hacdias replied:
Yeah, this wasn;t correct at all. Fixed and added to tests.

lidel commented on core/commands/name/name.go:
```diff
@@ -84,27 +79,25 @@ Resolve the value of a dnslink:
 
 type IpnsInspectResult struct {
 	Entry      IpnsInspectEntry
+	Version    string
```
Two asks here:

1.  This is present in `--enc=json`  output, but missing from default text one

2. Version may be confused with Sequence. Let' make it less ambigous, what matters is "signature type" used for validation, so maybe:
```suggestion
	SignatureType    string
```

This way we have "V1+V2" and "V2" as distinct signature types.

lidel commented on core/commands/name/name.go:
```diff
@@ -136,6 +129,7 @@ Passing --verify will verify signature against provided public key.
 	},
 	Options: []cmds.Option{
 		cmds.StringOption("verify", "CID of the public IPNS key to validate against."),
+		cmds.BoolOption("verbose", "Show a full hex dump of the raw Protobuf record."),
```
Realized we have two options starting with `ver`, which is pretty bad  for shell completion.
Let's change this to `--dump` and enable it by default:

```suggestion
		cmds.BoolOption("dump", "Include full hex dump of the raw Protobuf record.").WithDefault(true),
```

lidel commented on core/commands/name/name.go:
```diff
@@ -84,27 +79,25 @@ Resolve the value of a dnslink:
 }
 
 type IpnsInspectResult struct {
 	Entry      IpnsInspectEntry
```
Can we add protobuf size and print it before the hexdump and Version?
Will help with quick eyeballing and other analysis (V1 vs V2 vs records with additional data in Data CBOR)

```suggestion
	Entry      IpnsInspectEntry
	PbSize       int
```

lidel commented on test/cli/name_test.go:
Thank you for porting these :ok_hand:

lidel reviewed (changes requested):
Some UX papercuts and missing tests need to be addressed (details below),
and needs to be rebased after https://github.com/ipfs/boxo/pull/339 lands,
but otherwise lgtm.  

@hacdias do you think we should include this in 0.21  (https://github.com/ipfs/kubo/issues/9814), or do we do push both to 0.22? Since V2-only is opt-in, it should be safe to ship sooner, but ok to play it safe and push to 0.22

hacdias commented on core/commands/name/name.go:
```diff
@@ -231,24 +223,25 @@ Passing --verify will verify signature against provided public key.
+
 			if out.Entry.Validity != nil {
-				fmt.Fprintf(tw, "Validity:\t%s\n", out.Entry.Validity.Format(time.RFC3339Nano))
+				fmt.Fprintf(tw, "Validity:\t%q\n", out.Entry.Validity.Format(time.RFC3339))
```
No reason, reverted.

hacdias commented:
@lidel I addressed the feedback and merged on Boxo side. I will be merging here once CI passes.

Regarding Kubo 0.21: I don’t mind either way. If we do an RC4 (let’s hope not), I will include it. Otherwise, let’s move it to Kubo 0.22. In case we don’t want to include it in Kubo 0.21 and there’s an RC4, I do not mind doing a cherry-picked release.
